net/tstun/mtu.go

```diff
@@ -1,33 +1,171 @@
+	// in the absence of other information such as path MTU probes.
+	safeTunMTU TunMTU = 1280
 )
 
-// DefaultMTU returns either the constant default MTU of 1280, or the value set
-// in TS_DEBUG_MTU clamped to a maximum of 65536.
-func DefaultMTU() uint32 {
-	// DefaultMTU is the Tailscale default MTU for now.
-	//
-	// wireguard-go defaults to 1420 bytes, which only works if the
-	// "outer" MTU is 1500 bytes. This breaks on DSL connections
-	// (typically 1492 MTU) and on GCE (1460 MTU?!).
-	//
-	// 1280 is the smallest MTU allowed for IPv6, which is a sensible
-	// "probably works everywhere" setting until we develop proper PMTU
-	// discovery.
-	tunMTU := defaultMTU
-	if mtu, ok := envknob.LookupUintSized("TS_DEBUG_MTU", 10, 32); ok {
-		mtu := uint32(mtu)
-		if mtu > maxMTU {
-			mtu = maxMTU
+// WireMTUsToProbe are the MTUs we are most likely to see in real world
+// networks. These are used to send discovery pings in packets padded
+// to exactly match the goal MTU.
+//
+// XXX These need adjusting before merging to main. What do you think?
+var WireMTUsToProbe = [...]WireMTU{
+	576,  // Smallest MTU for IPv4, probably useless?
+	1124, // An observed MTU in the wild, maybe 1100 instead?
+	1280, // Smallest MTU allowed for IPv6, current default
+	1480, // A little less than most common MTU, for tunnels or such
+	1500, // Most common real world MTU
+	8000, // Some jumbo frames are this size
+	9000, // Most jumbo frames are this size or larger
+}
+
+// MaxProbedWireMTU is the largest MTU we will test for path MTU
+// discovery.
+var MaxProbedWireMTU WireMTU
+
+func init() {
+	for _, m := range WireMTUsToProbe {
+		// Remove MTUs larger than the device max MTU.
+		if m > WireMTU(maxTunMTU) {
+			// TODO: actually remove these from the list to probe.
+			continue
+		}
+		if m > MaxProbedWireMTU {
+			MaxProbedWireMTU = m
 		}
-		tunMTU = mtu
 	}
-	return tunMTU
+}
+
+// wgHeaderLen is the length of all the headers Wireguard adds to a packet
+// in the worst case (IPv6). This constant is for use when we can't or
+// shouldn't use information about the IP version of a specific packet
+// (e.g., calculating the MTU for the Tailscale interface.
+//
+// A Wireguard header includes:
+//
+// - 20-byte IPv4 header or 40-byte IPv6 header
+// - 8-byte UDP header
+// - 4-byte type
+// - 4-byte key index
+// - 8-byte nonce
+// - 16-byte authentication tag
+
+const wgHeaderLen = 40 + 8 + 4 + 4 + 8 + 16
+
+// TunToWireMTU converts the MTU that the Tailscale TUN presents to
+// the user to the on-the-wire MTU necessary to transmit a packet of
+// TUN MTU bytes plus the Tailscale/Wireguard overhead.
+func TunToWireMTU(t TunMTU) WireMTU {
+	return WireMTU(t + wgHeaderLen)
+}
+
+// WireToTunMTU converts the on-the-wire MTU to the Tailscale MTU
+// necessary to transmit a packet of TUN MTU bytes plus the
+// Tailscale/Wireguard overhead.
+func WireToTunMTU(w WireMTU) TunMTU {
+	if w < wgHeaderLen {
+		return 0
+	}
+	return TunMTU(w - wgHeaderLen)
```

Comment:
Thanks! This specific case was too hard to read using min() because the args are unsigned. So we can't do the straightforward max(w-wgHeaderLen, 0) and instead have to do w - min(w, wgHeaderLen) which ??? no thanks. But I used it in a bunch of other places.